File: clients/typescript/src/amount.ts
import type { SubscriptionDelegation } from '@solana/subscriptions';

import {
    CHANGE_MIN,
    CHANGE_STEP,
    DAY_SECONDS,
    INCOME_MIN,
    INCOME_SHARE_BPS,
    TIERS,
    TRIAL_CAP,
    TRIAL_SECONDS,
    WEEK_SECONDS,
} from './constants';
import { Engine, type MarketCalendar, type UserConfig, UserStatus } from './generated';
import { divEuclid, usMarketOpen } from './market';

const U64_MAX = 2n ** 64n - 1n;

/** What one sweep pulls, by source, in payment-token raw units. */
export type Pull = { engine: bigint; pending: bigint };

const min = (a: bigint, b: bigint) => (a < b ? a : b);
const saturatingSub = (a: bigint, b: bigint) => (a > b ? a - b : 0n);

/** The pull's total: exactly what the sweep takes from the user. */
export const pullTotal = (pull: Pull) => pull.engine + pull.pending;

/** The pull within `limit`, the engine kept first. */
export function cappedPull(pull: Pull, limit: bigint): Pull {
    const engine = min(pull.engine, limit);
    return { engine, pending: min(pull.pending, limit - engine) };
}

/** What the income rule adds for an incoming payment. */
export function incomeShare(user: UserConfig, income: bigint): bigint {
    return user.incomeRule && income >= INCOME_MIN ? (income * INCOME_SHARE_BPS) / 10_000n : 0n;
}

/** What change per payment adds for an outgoing payment: max(round-up to the next dollar, $0.50) × multiplier. */
export function change(user: UserConfig, payment: bigint): bigint {
    const roundUp = (CHANGE_STEP - (payment % CHANGE_STEP)) % CHANGE_STEP;
    return (roundUp > CHANGE_MIN ? roundUp : CHANGE_MIN) * BigInt(user.changeMultiplier);
}

/** Index of the user's week at `now`, counted from enrollment. */
export function weekAt(user: UserConfig, now: bigint): number {
    const elapsed = now - user.enrolledAt;
    return Number(BigInt.asUintN(32, (elapsed > 0n ? elapsed : 0n) / WEEK_SECONDS));
}

/** The combined cap across both payment tokens for the week of `now`, never above the beta's cap per user. */
export function weeklyCap(user: UserConfig, betaCap: bigint, now: bigint): bigint {
    const cap = now < user.enrolledAt + TRIAL_SECONDS ? TRIAL_CAP : (TIERS[user.tier as 0 | 1] ?? 0n);
    return min(cap, betaCap);
}

/** Pulled so far in the week of `now`, across both payment tokens. */
export function spentAt(user: UserConfig, now: bigint): bigint {
    return weekAt(user, now) === user.week ? user.weekSpent : 0n;
}

/** The daily engine buys once per UTC day, the weekly engine once per week during a regular NYSE session. */
export function engineDue(user: UserConfig, calendar: MarketCalendar, now: bigint): boolean {
    const ran = user.engineRanAt >= user.enrolledAt;
    if (user.engine === Engine.Daily) {
        return !ran || divEuclid(now, DAY_SECONDS) > divEuclid(user.engineRanAt, DAY_SECONDS);
    }
    return (!ran || weekAt(user, now) > weekAt(user, user.engineRanAt)) && usMarketOpen(now, calendar);
}

/**
 * What a sweep of `paymentToken` may pull at `now` from a `balance` in that token, as the program's `pull` computes
 * it: the engine if due, then pending amounts, within the week's remaining cap and above the token's cushion.
 */
export function pull(
    user: UserConfig,
    paymentToken: number,
    balance: bigint,
    betaCap: bigint,
    calendar: MarketCalendar,
    now: bigint,
): Pull {
    if (user.status !== UserStatus.Active) return { engine: 0n, pending: 0n };
    const cushion = user.cushions[paymentToken] ?? U64_MAX;
    const room = min(saturatingSub(weeklyCap(user, betaCap, now), spentAt(user, now)), saturatingSub(balance, cushion));
    const engine = engineDue(user, calendar, now) ? min(user.engineAmount, room) : 0n;
    return { engine, pending: min(user.pending, room - engine) };
}

/**
 * What a subscription still lets its plan's owner pull in the current period at `now`, as the program's
 * `subscription::remaining` bounds each sweep: 0 once it has expired, or when there is no subscription.
 */
export function nativeRemaining(subscription: SubscriptionDelegation | null, now: bigint): bigint {
    if (!subscription) return 0n;
    const { amount, periodHours } = subscription.terms;
    if (subscription.expiresAtTs !== 0n && now >= subscription.expiresAtTs) return 0n;
    const period = min(periodHours * 3_600n, U64_MAX);
    const elapsed = now - subscription.currentPeriodStartTs;
    if (elapsed >= 0n && elapsed >= period) return amount;
    return saturatingSub(amount, subscription.amountPulledInPeriod);
}

File: programs/laterite/src/amount.rs
//! The amount engine: how much a sweep may pull, derived only from the user's signed rules, the events attested for
//! them, the clock, the market calendar and the token balance. Amounts are payment-token raw units, one dollar being
//! 1,000,000.

use crate::{
    us_market_open, Engine, MarketCalendar, UserConfig, UserStatus, DAY_SECONDS, TIERS, TRIAL_CAP, TRIAL_SECONDS,
    WEEK_SECONDS,
};

/// Income rule: the share of each incoming payment invested, in basis points (10%).
pub const INCOME_SHARE_BPS: u64 = 1_000;

/// Income rule: smaller incoming payments are ignored ($50).
pub const INCOME_MIN: u64 = 50_000_000;

/// Change per payment rounds each outgoing payment up to the next dollar...
pub const CHANGE_STEP: u64 = 1_000_000;

/// ...and invests at least $0.50 per payment, before the multiplier.
pub const CHANGE_MIN: u64 = 500_000;

/// What one sweep pulls, by source.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct Pull {
    pub engine: u64,
    pub pending: u64,
}

impl Pull {
    pub fn total(&self) -> u64 {
        self.engine + self.pending
    }

    /// The pull within `limit`, the engine kept first.
    pub fn capped(self, limit: u64) -> Pull {
        let engine = self.engine.min(limit);
        Pull { engine, pending: self.pending.min(limit - engine) }
    }
}

impl UserConfig {
    /// What the income rule adds for an incoming payment.
    pub fn income_share(&self, income: u64) -> u64 {
        if !self.income_rule || income < INCOME_MIN {
            return 0;
        }
        (u128::from(income) * u128::from(INCOME_SHARE_BPS) / 10_000) as u64
    }

    /// What change per payment adds for an outgoing payment: max(round-up to the next dollar, $0.50) times the
    /// multiplier.
    pub fn change(&self, payment: u64) -> u64 {
        let round_up = (CHANGE_STEP - payment % CHANGE_STEP) % CHANGE_STEP;
        round_up.max(CHANGE_MIN) * u64::from(self.change_multiplier)
    }

    /// Index of the user's week at `now`, counted from enrollment.
    pub fn week_at(&self, now: i64) -> u32 {
        (now.saturating_sub(self.enrolled_at).max(0) / WEEK_SECONDS) as u32
    }

    /// The combined cap across both payment tokens for the week of `now`: the trial cap in the first week, the tier
    /// afterwards, never above the beta's cap per user.
    pub fn weekly_cap(&self, beta_cap: u64, now: i64) -> u64 {
        let cap = if now < self.enrolled_at.saturating_add(TRIAL_SECONDS) {
            TRIAL_CAP
        } else {
            TIERS.get(usize::from(self.tier)).copied().unwrap_or(0)
        };
        cap.min(beta_cap)
    }

    /// Pulled so far in the week of `now`, across both payment tokens.
    pub fn spent_at(&self, now: i64) -> u64 {
        if self.week_at(now) == self.week {
            self.week_spent
        } else {
            0
        }
    }

    /// The daily engine buys once per UTC day, at any hour; the weekly engine once per week, during a regular NYSE
    /// session. A missed day or week is skipped, not made up.
    pub fn engine_due(&self, calendar: &MarketCalendar, now: i64) -> bool {
        let ran = self.engine_ran_at >= self.enrolled_at;
        match self.engine {
            Engine::Daily => !ran || now.div_euclid(DAY_SECONDS) > self.engine_ran_at.div_euclid(DAY_SECONDS),
            Engine::Weekly => {
                (!ran || self.week_at(now) > self.week_at(self.engine_ran_at)) && us_market_open(now, calendar)
            }
        }
    }

    /// What a sweep of `payment_token` may pull at `now` from a `balance` in that token: the engine if due, then
    /// pending variable amounts, within the week's remaining cap and above the token's cushion. Nothing unless the
    /// user is active.
    pub fn pull(&self, payment_token: usize, balance: u64, beta_cap: u64, calendar: &MarketCalendar, now: i64) -> Pull {
        if self.status != UserStatus::Active {
            return Pull::default();
        }
        let cushion = self.cushions.get(payment_token).copied().unwrap_or(u64::MAX);
        let room =
            self.weekly_cap(beta_cap, now).saturating_sub(self.spent_at(now)).min(balance.saturating_sub(cushion));
        let engine = if self.engine_due(calendar, now) { self.engine_amount.min(room) } else { 0 };
        Pull { engine, pending: self.pending.min(room - engine) }
    }

    /// Records a completed pull: rolls the week over, counts the pull against it, consumes the pending amount and
    /// marks the engine as run.
    pub fn record(&mut self, pull: Pull, now: i64) {
        let week = self.week_at(now);
        if week != self.week {
            self.week = week;
            self.week_spent = 0;
        }
        self.week_spent = self.week_spent.saturating_add(pull.total());
        self.pending = self.pending.saturating_sub(pull.pending);
        if pull.engine > 0 {
            self.engine_ran_at = now;
        }
    }
}
